addons/website/models/website_visitor.py

```diff
@@ -161,88 +181,113 @@ def action_send_mail(self):
+                            AND visit_datetime >= NOW() AT TIME ZONE 'UTC' - INTERVAL '30 minutes'
+                        )
+                    )
+                    SELECT id from visitor
+                """
+                query_values += tuple(track_values.values()) + tuple(where_params)
 
-        if not self.env.user._is_public():
-            partner_id = self.env.user.partner_id
-            if not visitor or visitor.partner_id and visitor.partner_id != partner_id:
-                # Partner and no cookie or wrong cookie
-                visitor = Visitor.with_context(active_test=False).search([('partner_id', '=', partner_id.id)])
-        elif visitor and visitor.partner_id:
-            # Cookie associated to a Partner
-            visitor = Visitor
-
-        # also check that visitor parent partner is not different from user's one
-        # (indicates duplicate due to invalid or wrong cookie)
-        if visitor and visitor.parent_id.partner_id:
-            if self.env.user._is_public():
-                visitor = self.env['website.visitor'].sudo()
-            elif not visitor.partner_id:
-                visitor = self.env['website.visitor'].sudo().with_context(active_test=False).search(
-                    [('partner_id', '=', self.env.user.partner_id.id)]
-                )
-
-        if visitor and not visitor.timezone:
+            self.env.cr.execute(query, query_values)
+            visitor_id = self.env.cr.fetchone()[0]
+            visitor = Visitor.browse(visitor_id)
+            visitor._post_create()
+        else:
+            visitor = Visitor.search([('access_token', '=', access_token)])
+
+        if not force_create and visitor and not visitor.timezone:
             tz = self._get_visitor_timezone()
             if tz:
                 visitor._update_visitor_timezone(tz)
 
-        if not visitor and force_create:
-            visitor = self._create_visitor()
-
         return visitor
 
-    def _handle_webpage_dispatch(self, response, website_page):
-        # get visitor. Done here to avoid having to do it multiple times in case of override.
-        visitor_sudo = self._get_visitor_from_request(force_create=True)
-        if request.httprequest.cookies.get('visitor_uuid', '') != visitor_sudo.access_token:
-            expiration_date = datetime.now() + timedelta(days=365)
-            response.set_cookie('visitor_uuid', visitor_sudo.access_token, expires=expiration_date)
-        self._handle_website_page_visit(website_page, visitor_sudo)
-
-    def _handle_website_page_visit(self, website_page, visitor_sudo):
-        """ Called on dispatch. This will create a website.visitor if the http request object
-        is a tracked website page or a tracked view. Only on tracked elements to avoid having
-        too much operations done on every page or other http requests.
-        Note: The side effect is that the last_connection_datetime is updated ONLY on tracked elements."""
+    def _post_create(self):
```

Comment:
This 'void placeholder with a todo' indicates you have an issue with the update of your model.

Reply:
here's an idea: extract the visitor upsert out of `_get_visitor_from_request()`, call it `_upsert_visitor()`, and have `website_livechat` override it to plug its stuff after calling super().